keps/sig-node/20200110-apparmor-ga.md

```diff
@@ -0,0 +1,390 @@
+
+AppArmor support has been added with Kubernetes v1.14 and is already in beta.
+Profiles have to be available on each node whereas the container runtime ensures
+that the profile is loaded when specified at pod or PSP level. Profiles
+can be specified per-container via the pod’s metadata annotation:
+
+```
+container.apparmor.security.beta.kubernetes.io/<container_name>: {unconfined,runtime/default,localhost/<profile>}
+```
+
+The feature has been more or less unchanged ever since. Also note that the
+addition predates feature gates or our modern concept of feature lifecycle. So,
+even though the annotations include `beta` in the key, this is entirely useable
+on any production GA cluster.
+
+The main motivation behind this KEP is to promote the AppArmor feature to GA.
+
+_NOTE: Seccomp is in a very similar state, but with some subtle differences.
+Promoting Seccomp to GA will be covered by a [separate
+KEP][https://github.com/kubernetes/enhancements/pull/1148]._
+
+### Goals
+
+- Declare AppArmor as GA
+- Fully document and formally spec the feature support
+- Add equivalent API fields to replace AppArmor annotations
+- Deprecate the AppArmor annotations
+
+### Non-Goals
+
+This KEP proposes the absolute minimum to get AppArmor to GA, therefore all
+functional enhancements are out of scope, including:
+
+- Changing the default AppArmor profile from `unconfined`
+- Defining any standard "Kubernetes branded" AppArmor profiles
+- Formally specifying the AppArmor profile format in Kubernetes
+- Providing mechanisms for loading profiles from outside the of the node
+- Changing the semantics around AppArmor support
+- Windows support
+
+## Proposal
+
+AppArmor is not available on every Linux distribution. Beside this, container
+runtimes have AppArmor as compile-time feature which may be disabled as well.
+With the GA API we do not change the error handling and behave exactly the same
+as the current error propagation paths.
+
+### API
+
+The AppArmor API will be functionally equivalent to the current beta API. This
+includes the Pod API, which specifies what profile the containers run with, and
+the `PodSecurityPolicy` API which specifies allowed profiles & a default
+profile.
+
+#### Pod API
+
+The Pod AppArmor API is generally immutable.
+
+```go
+type PodSecurityContext struct {
+    ...
+    // The AppArmor options to use by the containers in this pod.
+    // +optional
+    AppArmor  *AppArmorOptions
+    ...
+}
+
+type SecurityContext struct {
+    ...
+    // The AppArmor options to use by this container.
+    // +optional
+    AppArmor  *AppArmorOptions
+    ...
+}
+
+type AppArmorOptions struct {
+    // The AppArmor profile to run with.
+    Profile
+}
+
+// Only one profile source may be set.
+// +union
+type AppArmorProfile struct {
+    // +unionDescriminator
+    Type AppArmorProfileType
+
+    // Use a predefined profile defined by the runtime.
+    // Most runtimes only support "default"
+    // +optional
+    RuntimeProfile *string
```

Comment:
I'm still torn on whether we should include this field initially. Currently there aren't any other runtime profiles accepted, and it makes reconciliation with the annotation more complicated.

I'm leaning towards dropping this field, and just having the type `Runtime` be synonymous with `runtime/default`

Reply:
Alright, I removed the field and renamed the `Runtime` type to `Default`.